### ISSUE TITLE
tf.distribute.MirroredStrategy hangs when custom op used NCCL

### ISSUE THREAD

**Issue by Jianbing-D** [tensorflow/tensorflow#48753]
TensorFLow version: 2.4.1
python version: 3.8.5
CUDA version: 11.2
CUDA Driver version: 460.32.03
NCCL version: 2.8.4

Issue elaboration:
I implemented a custom op where NCCL API are used to exchange data among GPUs. And I used this custom op with `tf.distribute.MirroredStrategy`.
Here is the status:

- Situation 1:

```python
class CustomLayer(tf.keras.layers.Layer):
    ...
    def call(inputs, training=True):
        outputs = custom_op(inputs, training=training)
        return outputs

strategy = tf.distribute.MirroredStrategy()

with strategy.scope():
     layer = CustomLayer()

@tf.function
def _step(input):
    return layer(input)

for iterations in range(10):
    output = strategy.run(_step, args=(inputs,)) # where inputs are an PerReplica data
```
I used 8 GPUs and set `NCCL_LAUNCH_MODE=PARALLEL` to run the above code snippet, **it works.**

- Situation 2:
```python
class CustomLayer(tf.keras.layers.Layer):
    ...
    def call(inputs, training=True):
        outputs = custom_op(inputs, training=training)
        return outputs

strategy = tf.distribute.MirroredStrategy()

with strategy.scope():
     layer = CustomLayer()
     var = tf.Variable(initial_value=1.0, dtype=tf.float32)

@tf.function
def _step(input):
    return var * layer(input)

for iterations in range(10):
    output = strategy.run(_step, args=(inputs,)) # where inputs are an PerReplica data
```
I still used 8 GPUs and set `NCCL_LAUNCH_MODE=PARALLEL`  to run the above code snippet. But only 1 or 2 iterations will be executed successfully, then the program will **hang**.  <br>
I tried to add `std::cout` inside of my custom op to check which GPU's OpKernel is not executed. And it always shows that `GPU 0` 's OpKernel::Compute() is not executed. And other GPUs are waiting for GPU0 to arrive at NCCL API calling point.

- Situation 3:
```python
class CustomLayer(tf.keras.layers.Layer):
    ...
    def call(inputs, training=True):
        outputs = custom_op(inputs, training=training)
        return outputs

strategy = tf.distribute.MirroredStrategy()

with strategy.scope():
     layer = CustomLayer()
     var = tf.Variable(initial_value=1.0, dtype=tf.float32)

@tf.function
def _step(input):
    return var * layer(input)

for iterations in range(10):
    output = strategy.run(_step, args=(inputs,)) # where inputs are an PerReplica data
```
The same code with situation 2, and 8 GPUs are used. But **not set** environment variable `NCCL_LAUNCH_MODE=PARALLEL`. And it **works**, all iterations are successfully executed.
<br>

**My question is:**
How can I find out why `GPU0` hangs after 1 or 2 iterations in situation 2?  <br>
How to find out which operation caused TensorFlow stucks for executing the Opkernel::Comput()?

**Comment by Jianbing-D** [comment]
If you want to reproduce this issue, here are the steps:

1. download files from https://drive.google.com/file/d/1VYqAbJHM_DT2wtVZ4RD5PW1vWNr76hz1/view?usp=sharing
2. extract source files from downloaded files, and all the files I mentioned below can be found here.
3. Build docker image using "cuda11_tf2_ubuntu20.04.dockerfile" and start a docker container with that image
4. enter that docker container with interactive mode
5. mkdir -p build
6. (Optional) if your GPU's compute capability is not 70, you should specify the correct "sm" in "compile.sh"
7. bash compile.sh, after this command, an dynamic lib will be compiled and installed to your /usr/local/lib
8. python main.py, when the program will be hanging forever.

**Comment by Saduf2019** [comment]
@Jianbing-D 
Can you please let us know if this si still an issue in latest tf version.

**Comment by Jianbing-D** [comment]
I think it is only related to my customized codes. And I am debugging it.

**Comment by marsmiao** [comment]
> I think it is only related to my customized codes. And I am debugging it.

Have you got anything yet? I'm facing the same question while using the code you wrote (Sparse Operation Kit), I think it maybe caused by Gradient Tape

**Comment by Jianbing-D** [comment]
> > I think it is only related to my customized codes. And I am debugging it.
> 
> Have you got anything yet? I'm facing the same question while using the code you wrote (Sparse Operation Kit), I think it maybe caused by Gradient Tape

Could you please provide more info about your issue, and directly ask it in this 【[repo](https://github.com/NVIDIA-Merlin/HugeCTR)】?

**Comment by Venkat6871** [comment]
Hi **@Jianbing-D** ,
Can you please let us know if this is still an issue in latest tf version ? Please feel free to close the issue if it is resolved .

Thank you!
